include/swift/AST/DiagnosticConsumer.h

```diff
@@ -110,6 +111,19 @@ class DiagnosticConsumer {
   /// so here's a placeholder.
 
   virtual void informDriverOfIncompleteBatchModeCompilation() {}
+  
+private:
+  /// Some subclasses need to track if handleDiagnostic has ever seen an error
+  bool hasAnErrorBeenHandled = false;
```

Comment:
I don't think it is net-adding code:
Of the 14 subclasses of DiagnosticConsumer 5 (EditorDiagnosticConsumer, ErrorGatherer, FileSpecificDiagnosticConsumer, FormatterDiagnosticConsumer, PrintingDiagnosticConsumer, and SerializedDiagnosticConsumer) all need to track errors. For each of these, two lines of code are saved: the boolean declaration, and the accessor function. That's 10 lines saved.

The other 9 subclasses gain one line to call the error-tracking function. That's 9 lines added.

So, it is a net savings of 1 line of code.

Yes, it is something most consumers don't need. I do agree that my current design does not centralize the invariant about error-tracking, and that is a problem.

I think I may have an alternative that's better: instead of having each client directly call a `handleDiagnostic` function in each subclass, how about if each client calls a public `handleDiagnostic` function defined only in `DiagnosticConsumer`. That function can do the error-tracking and call a protected `finishHandleDiagnostic` function. This latter function would be abstract in `DiagnosticConsumer` and implemented in each subclass to do what `handleDiagnostic` does today. I'll see if I can do that this morning.